---
title: "script_2.1_process_20250805"
author: "GKM"
date: "2025-08-11"
output:
  html_document:
    keep_md: TRUE
---

This script is a pipeline to process with 20250805_run1 samples. It will explore the seurat file, plot and save various plots and determine the QC metrics for both PBMC and T cell sorted populations. 
However, to save the processed seurat object, should run as a batch script, found at ~/R/scCHANTS/script_2.11_save_processed_20250805_pbmc_rds.R or ~/R/scCHANTS/script_2.11_save_processed_20250805_t_rds.R

* unzip files
* load into a seurat object
* (in future scripts will need to do data integration)
* filter and quality control
* normalisation
* (batch effect correction)
* feature selection
* dimensionality reduction
* export Rds file

Before running bear in mind all plots will automatically be saved, I recommend copying plots to your OneDrive folder as you generate them in case of overwriting (though they should be the same!)
Can also search for " ggsave" and replace with "#ggsave" or vice versa

# Script 2.1a Load into R and HTO demultiplex

## Load packages
```{r}
library("R.utils")    # need for decompressing files
library("dplyr")
library("Seurat")
library("ggplot2")
library("clustree")
library("patchwork")
library("viridis")
```

## Create file path for saving plots
Must create this directory and sub-directories e.g. demultiplex, QC, PCA
```{r}
file_path <- "/scratch/prj/id_hill_sims_wellcda/scCHANTS/20250805_run1/plots"
```

# Donor 55
## Load data & rename HTO and ADTs
```{r}
## Decompress the files

# only need to do once
# gunzip("/scratch/prj/id_hill_sims_wellcda/scCHANTS/20250805_run1/20250805_run1_055/matrix.mtx.gz", temporary = TRUE, remove = FALSE)
# gunzip("/scratch/prj/id_hill_sims_wellcda/scCHANTS/20250805_run1/20250805_run1_055/barcodes.tsv.gz", temporary = TRUE, remove = FALSE)
# gunzip("/scratch/prj/id_hill_sims_wellcda/scCHANTS/20250805_run1/20250805_run1_055/features.tsv.gz", temporary = TRUE, remove = FALSE)


## Load data

# load the unzipped files
# gene.column = 2 gives gene symbol/name instead of ENSEMBL ID
scCHANTS_055_data <- Read10X(data.dir = "/scratch/prj/id_hill_sims_wellcda/scCHANTS/20250805_run1/20250805_run1_055/", gene.column=2)

# create seurat object
# could add arguments for min.cells or min.features to be included
scCHANTS_055 <- CreateSeuratObject(counts = scCHANTS_055_data, project = "scCHANTS_055")

Assays(scCHANTS_055)
Layers(scCHANTS_055[["RNA"]])


## Rename HTO and ADT ids

# add HTO and ADT assay to the seurat object as it is by default ignored by CreateSeuratObject
# extract feature names
HTO_names <- rownames(scCHANTS_055_data$`Antibody Capture`) [1:13]
ADT_names <- rownames(scCHANTS_055_data$`Antibody Capture`) [14:17]

# extract count matrices
HTO_counts <- scCHANTS_055_data$`Antibody Capture` [HTO_names, , drop=FALSE]
ADT_counts <- scCHANTS_055_data$`Antibody Capture` [ADT_names, , drop=FALSE]

# check names and order
rownames(HTO_counts)
rownames(ADT_counts)

# create mapping from long to short names. using - not _ as cannot have underscores in seurat object names
name_to_id <- c(
  "TotalSeq™-C0251 anti-human Hashtag 1 Antibody"  = "HTO-1",
  "TotalSeq™-C0252 anti-human Hashtag 2 Antibody"  = "HTO-2",
  "TotalSeq™-C0253 anti-human Hashtag 3 Antibody"  = "HTO-3",
  "TotalSeq™-C0254 anti-human Hashtag 4 Antibody"  = "HTO-4",
  "TotalSeq™-C0255 anti-human Hashtag 5 Antibody"  = "HTO-5",
  "TotalSeq™-C0256 anti-human Hashtag 6 Antibody"  = "HTO-6",
  "TotalSeq™-C0257 anti-human Hashtag 7 Antibody"  = "HTO-7",
  "TotalSeq™-C0258 anti-human Hashtag 8 Antibody"  = "HTO-8",
  "TotalSeq™-C0259 anti-human Hashtag 9 Antibody"  = "HTO-9",
  "TotalSeq™-C0260 anti-human Hashtag 10 Antibody" = "HTO-10",
  "TotalSeq™-C0296 anti-human Hashtag 11 Antibody" = "HTO-11",
  "TotalSeq™-C0262 anti-human Hashtag 12 Antibody" = "HTO-12",
  "TotalSeq™-C0263 anti-human Hashtag 13 Antibody" = "HTO-13",
  
  # ADTs
  "TotalSeq-C anti human CD14 (clone M5E2, cat 301859)"             = "ADT-CD14",
  "TotalSeq-C anti human CD16 (clone 3G8 cat 302065)"               = "ADT-CD16",
  "TotalSeq™-C0045 anti-human CD4 Antibody (clone SK3, cat 344651)" = "ADT-CD4",
  "TotalSeq™-C0046 anti-human CD8 Antibody (clone SK1, cat 344753)" = "ADT-CD8"
)

# rename the rows (features) using the mapping
rownames(HTO_counts) <- unname(name_to_id[rownames(HTO_counts)])
rownames(ADT_counts) <- unname(name_to_id[rownames(ADT_counts)])

# check names have changed
rownames(HTO_counts)
rownames(ADT_counts)


## Add HTO and ADT assays to seurat object

# add HTO assay
scCHANTS_055[["HTO"]] <- CreateAssayObject(counts = HTO_counts)

# add ADT assay
scCHANTS_055[["ADT"]] <- CreateAssayObject(counts = ADT_counts)

# validate assays were added
Assays(scCHANTS_055)

# validate layers
Layers(scCHANTS_055)

# check default assay
DefaultAssay(scCHANTS_055)

# if needed, can change default assay
DefaultAssay(scCHANTS_055) <- "RNA"

# check number of cells (samples) and features looks right
scCHANTS_055@assays$RNA
scCHANTS_055@assays$HTO
scCHANTS_055@assays$ADT
```


```{r}
# free space
rm(scCHANTS_055_data, ADT_counts, HTO_counts, ADT_names, HTO_names, name_to_id)
```

## Demultiplex HTOs

Should check in whether to normalise across features or cells (use margin argument)

https://satijalab.org/seurat/articles/hashing_vignette

HTODemux:

* Perform a k-medoid clustering on the normalized HTO values, which initially separates cells into K(# of samples)+1 clusters.
* Calculate a ‘negative’ distribution for HTO. For each HTO, the cluster with the lowest average value is used as the negative group.
* For each HTO, fit a negative binomial distribution to the negative cluster. 0.99 quantile of this distribution is used as a threshold.
* Based on these thresholds, each cell is classified as positive or negative for each HTO.
* Cells that are positive for more than one HTOs are annotated as doublets.

```{r}
# remove empty
scCHANTS_055$totRNA <- Matrix::colSums(scCHANTS_055@assays$RNA@layers$`counts.Gene Expression`)
scCHANTS_055$totHTO <- Matrix::colSums(scCHANTS_055@assays$HTO@counts)

FeatureScatter(scCHANTS_055, "totRNA", "totHTO") + ggtitle("RNA vs HTO per barcode")
VlnPlot(scCHANTS_055, features = c("totRNA","totHTO"), ncol = 2)

scCHANTS_055 <- subset(scCHANTS_055, subset = totRNA > 0 | totHTO > 0)

DefaultAssay(scCHANTS_055) <- "HTO"

# normalise HTO counts, standard is centered log-ratio (CLR) transformation
# can normalise across features (margin = 1) or across cells (margin = 2)
# comment on github forum satija lab says they recommend normalising across tags if variation between hash performances but since have very low signals, will try across cells
scCHANTS_055 <- NormalizeData(scCHANTS_055, assay = "HTO", normalization.method = "CLR", margin = 2)

# using default threshold (quantile of inferred negative distribution for each hashtag), default kfunc is clara (for fast k-medoids clustering on large applications, it is faster and more memory-efficient as it repeatedly samples subsets and clusters those)
scCHANTS_055 <- HTODemux(scCHANTS_055, assay = "HTO", positive.quantile = 0.99, kfunc = "clara")


## Visualise demultiplexing

DefaultAssay(scCHANTS_055) <- "HTO"

# look at classification
table(scCHANTS_055$HTO_classification.global)
## check most are singlets
```

```{r fig.dim = c(9,9)}
# visualise enrichment for some HTOs, group cells based on the max HTO signal
d1 <- RidgePlot(scCHANTS_055, assay = "HTO", features = rownames(scCHANTS_055[["HTO"]])[1:4], ncol = 2, group.by = "HTO_maxID")
d1
# ggsave("HTO_1to4_ridgeplot_055.png", plot = d1, device = png, height = 25, width = 30, unit = "cm", path = paste0(file_path, "/demultiplex_055/") )
```

```{r}
# look at pairs to see if mutually exclusive
d2 <- FeatureScatter(scCHANTS_055, feature1 = "HTO-1", feature2 = "HTO-2")
d2
# ggsave("HTO_1vs2_featureplot_055.png", plot = d2, device = png, height = 10, width = 15, unit = "cm", bg = "white", path = paste0(file_path, "/demultiplex_055/") )

# compare number UMIs
d3 <- VlnPlot(scCHANTS_055, features = "nCount_RNA", pt.size = 0.1, log = TRUE, group.by = "HTO_classification.global")
d3
# ggsave("HTO_nCount_RNA_055.png", plot = d3, device = png, height = 10, width = 18, unit = "cm", path = paste0(file_path, "/demultiplex_055/") )


# first remove negative cells from the object
scCHANTS_nonegs <- subset(scCHANTS_055, subset = HTO_classification.global != "Negative")
DefaultAssay(scCHANTS_055) <- "RNA"

# should check numbers of cells after subsetting as does not always subset every assay
length(Cells(scCHANTS_055))
length(Cells(scCHANTS_nonegs))
length(Cells(scCHANTS_nonegs[["RNA"]]))
length(Cells(scCHANTS_nonegs[["HTO"]]))
length(Cells(scCHANTS_nonegs[["ADT"]]))
## e.g. if counts.Gene Expression and counts.Antibody Capture are removed then RNA layer is no longer subsetted

# calculate a tSNE embedding of the HTO data
DefaultAssay(scCHANTS_nonegs) <- "HTO"

scCHANTS_nonegs <- ScaleData(scCHANTS_nonegs, features = rownames(scCHANTS_nonegs), verbose = TRUE)

scCHANTS_nonegs <- RunPCA(scCHANTS_nonegs, features = rownames(scCHANTS_nonegs), approx = FALSE)

scCHANTS_nonegs <- RunUMAP(scCHANTS_nonegs, dims = 1:8, perplexity = 100)

table(Idents(scCHANTS_nonegs))

d4 <- DimPlot(scCHANTS_nonegs, group.by = "hash.ID")
d4
# ggsave("HTO_hashID_umap_055.png", plot = d4, device = png, height = 10, width = 15, unit = "cm", path = paste0(file_path, "/demultiplex_055/") )

d5 <- DimPlot(scCHANTS_nonegs, group.by = "HTO_classification.global")
d5
# ggsave("HTO_classification_umap_055.png", plot = d5, device = png, height = 10, width = 15, unit = "cm", path = paste0(file_path, "/demultiplex_055/") )

```

```{r fig.dim = c(8,8)}
## Create a heatmap of HTO signals

d6 <- HTOHeatmap(scCHANTS_055, assay = "HTO", ncells = 5000)
d6
# ggsave("HTO_signal_heatmap_055.png", plot = d6, device = png, height = 10, width = 15, unit = "cm", path = paste0(file_path, "/demultiplex_055/") )
```

```{r}
## Visualise RNA clustering (just mock to check for batch effects, will do clustering etc later on after QC)

# extract the singlets
scCHANTS_singlets <- subset(scCHANTS_055, HTO_classification.global = "Singlet")

DefaultAssay(scCHANTS_singlets) <- "RNA"
DefaultLayer(scCHANTS_singlets[["RNA"]])
# select the top 1000 most variable features
scCHANTS_singlets <- FindVariableFeatures(scCHANTS_singlets, assay = "RNA", selection.method = "vst", layer = "counts.Gene Expression")

# normalise and scale RNA data, only scale the variable features here for efficiency
scCHANTS_singlets <- NormalizeData(scCHANTS_singlets, assay = "RNA", layer = "counts.Gene Expression")
scCHANTS_singlets <- ScaleData(scCHANTS_singlets, features = VariableFeatures(scCHANTS_singlets))

# run PCA
scCHANTS_singlets <- RunPCA(scCHANTS_singlets, features = VariableFeatures(scCHANTS_singlets))

# select the top 10 PCs for clustering and tSNE based on PCElbowPlot
scCHANTS_singlets <- FindNeighbors(scCHANTS_singlets, reduction = "pca", dims = 1:10)
scCHANTS_singlets <- FindClusters(scCHANTS_singlets, resolution = 0.6, verbose = FALSE)
scCHANTS_singlets <- RunUMAP(scCHANTS_singlets, reduction = "pca", dims = 1:10)

# project singlet identities on UMAP visualisation
d7 <- DimPlot(scCHANTS_singlets, group.by = "HTO_classification.global") +
  labs(title = "RNA UMAP HTO_classification 055.global")
d7
# ggsave("HTO_classification_RNA_UMAP_055.png", plot = d7, device = png, height = 10, width = 15, unit = "cm", path = paste0(file_path, "/demultiplex_055/") )

# show only a subset
subset1 <- subset(scCHANTS_singlets, subset = HTO_maxID %in% c("HTO-1", "HTO-2", "HTO-3") )
d8 <- DimPlot(subset1, group.by = "HTO_maxID") +
  labs(title = "RNA UMAP HTO_maxID on HTO-1:3 055")
d8
# ggsave("HTO_1to3_maxID_RNA_UMAP_055.png", plot = d8, device = png, height = 10, width = 15, unit = "cm", path = paste0(file_path, "/demultiplex_055/") )
```

```{r fig.dim = c(11, 5)}
d9 <- DimPlot(subset1, group.by = "HTO_classification") +
  labs(title = "RNA UMAP HTO_classification on HTO-1:3") +
  theme(legend.text=element_text(size=6))
d9
# ggsave("HTO_1to3_classification_RNA_UMAP.png", plot = d9, device = png, height = 10, width = 20, unit = "cm", path = paste0(file_path, "/demultiplex/") )

subset2 <- subset(scCHANTS_singlets, subset = HTO_maxID %in% c("HTO-7", "HTO-8", "HTO-9") )
DimPlot(subset2, group.by = "HTO_classification")

```


## Check HTO data
20250616 data, shallow sequencing has lead to low HTO signal, check the signal here
```{r}
# check HTO counts
HTO_counts <- GetAssayData(scCHANTS, assay = "HTO", layer = "counts")
HTO_counts <- data.frame(HTO_UMIs = Matrix::colSums(HTO_counts))
summary(Matrix::colSums(HTO_counts))
## shows UMI counts are not too low

## checked HTOs are labelled as antibody capture in cell ranger
# check library size
d10 <- ggplot(HTO_counts, aes(x = HTO_UMIs)) +
  geom_histogram(bins = 150, fill = "steelblue", color = "black") +
  labs(main = "Total HTO UMIs per Cell", xlab = "HTO UMIs") +
  theme_minimal()
d10
## not peaking around 0 which is good
# ggsave("HTO_umis_per_cell_histogram.png", plot = d10, device = png, height = 7, width = 12, unit = "cm", path = paste0(file_path, "/demultiplex/") )


```


# Reassign scCHANTS to only singlets
```{r}
scCHANTS_full <- scCHANTS
length(Cells(scCHANTS))

scCHANTS <- subset(scCHANTS, subset = HTO_classification.global == "Singlet")

# check cell numbers
length(Cells(scCHANTS))
length(Cells(scCHANTS[["RNA"]]))
length(Cells(scCHANTS[["HTO"]]))
length(Cells(scCHANTS[["ADT"]]))
## not subsetting RNA assay
```

